src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
+  AstExpr() : ptr{} {
+  }
+  AstExpr(AstNodePtr ptr) : ptr{move(ptr)} {
+  }
+  AstExpr(const AstExpr& other) {
+    ptr = move(other.ptr);
+  }
+  AstExpr& operator=(const AstExpr& other) {
+    ptr = move(other.ptr);
+    return *this;
+  }
+  operator AstNodePtr() {
+    return move(ptr);
```

Comment:
Are you sure you want to `move(ptr)` here? This will `null`ify it internally, and the method's signature doesn't convey this IMHO.
What is the intended use of this method? If you want to provide non-destroying access, I'd return a `const&` to `ptr`. If you do intend to destroy the object, I'd make this very explicit from the method's name (definitely not an implicit conversion operator), and perhaps even make this an rvalue-only method (i.e. `AstNode DestroyingGet() && { ... }`)